migrating the take() fn from TH to ATen

bdhirsh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45430 migrate cuda implementation of take() from TH to ATen
* **#45283 migrating the take() fn from TH to ATen**

Differential Revision: [D24037298](https://our.internmc.facebook.com/intern/diff/D24037298)

bdhirsh committed:
migrating the take() fn from TH to ATen

[ghstack-poisoned]

bdhirsh committed:
Update on "migrating the take() fn from TH to ATen"




[ghstack-poisoned]

bdhirsh committed:
Update on "migrating the take() fn from TH to ATen"




[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5375,14 +5375,14 @@
+    CPU: take_cpu
     CUDA: legacy::cuda::_th_take
 
 - func: take_backward(Tensor grad, Tensor input, Tensor index) -> Tensor
```
I don't see anything requiring migration for take_backward. It's already defined in ATen/native here: https://github.com/pytorch/pytorch/blob/76ee58e2ec48d6922c49d647b276c1575a96aa09/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L818

bdhirsh commented on aten/src/ATen/native/take.cpp:
```diff
@@ -0,0 +1,105 @@
+#include <ATen/ATen.h>
```
There's probably a better name for this file? The file with the legacy definition just lumped together all of the remaining legacy function definitions. Or I could move everything here into TensorAdvancedIndexing.cpp where `take_backward()` lives.

bdhirsh committed:
Update on "migrating the take() fn from TH to ATen"




[ghstack-poisoned]

bdhirsh committed:
Update on "migrating the take() fn from TH to ATen"




[ghstack-poisoned]

mrshenli commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -815,6 +835,73 @@ Tensor masked_select_backward(const Tensor& grad, const Tensor& input, const Ten
+    bool is_contiguous = input.is_contiguous();
+    auto input_size = input.numel();
+
+    AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Bool, input.scalar_type(), "take_cpu", [&] {
```
Curious, do you know why do we special case `ScalarType::Bool` here instead hard-code it in `AT_DISPATCH_ALL_TYPES_AND` macro?

bdhirsh replied:
Only what I gathered from the comment here, which is that Tensors with bool dtype aren't a very common use case for most operators. I'm not really sure why `take` belongs to that category, but in the spirit of migration I figured I'd keep the support in. https://github.com/pytorch/pytorch/blob/181afd52205d974ff99d30dfa3d3033a71a47e2e/aten/src/ATen/Dispatch.h#L122

mrshenli commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -815,6 +835,73 @@ Tensor masked_select_backward(const Tensor& grad, const Tensor& input, const Ten
+                if (idx < input_size && idx >= -input_size) {
+                    idx = wrapLinearIndex(idx, input_size);
+                    if (is_contiguous) {
+                        output_data[i] = input_data[idx];
```
This requires input to be 1d instead of contiguous? See the following code. 

```python
>>> import torch
>>> x = torch.zeros(2, 2)
>>> x.is_contiguous()
True
>>> y = x[:, :2]
>>> y.is_contiguous()
False
```

mrshenli replied:
Please ignore the above comment, just realized you are using the data pointer instead of the tensor itself.

mrshenli commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -815,6 +835,73 @@ Tensor masked_select_backward(const Tensor& grad, const Tensor& input, const Ten
+        });
+
+        if (invalidIdxPos >= 0) {
+            checkLinearIndex(index_data[invalidIdxPos], input_size);
```
I wonder if it is possible to avoid an atomic var and just use a boolean flag to mark succeeded or failed. If it failed, we can then go through all indices to extra the culprit. There will still be contention on that flag. But if we set it to false (succeeded) before entering the loop, and then only set it true when it fails in the parallel loop, will it work?

mrshenli reviewed (approved):
LGTM! Left some minor comments.

bdhirsh committed:
Update on "migrating the take() fn from TH to ATen"




[ghstack-poisoned]

bdhirsh committed:
Update on "migrating the take() fn from TH to ATen"




[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -135,6 +135,26 @@ static Tensor reshape_indexer(const Tensor& index, int64_t dims_before, int64_t
+}
+
+static inline void checkLinearIndex(int64_t linearIndex, int64_t numel) {
+  TORCH_CHECK(linearIndex < numel && linearIndex >= -numel, "out of range: %d out of %d", (int)linearIndex, (int)numel);
```
I think I just pulled it from another file- agreed that the format string isn't as nice. I'll fix that

bdhirsh committed:
Update on "migrating the take() fn from TH to ATen"




[ghstack-poisoned]

bdhirsh committed:
Update on "migrating the take() fn from TH to ATen"




[ghstack-poisoned]

bdhirsh committed:
Update on "migrating the take() fn from TH to ATen"


Differential Revision: [D24037298](https://our.internmc.facebook.com/intern/diff/D24037298)

[ghstack-poisoned]
